kusto/ingest/ingest.go

```diff
@@ -60,16 +60,31 @@ type Ingestion struct {
 
 	connMu     sync.Mutex
 	streamConn *conn.Conn
+
+	blockSize   int
+	concurrency int
 }
 
-// New is the constructor for Ingestion.
-func New(client *kusto.Client, db, table string) (*Ingestion, error) {
-	mgr, err := getManager(client)
-	if err != nil {
-		return nil, err
+// Option is an optional argument to New().
+type Option func(s *Ingestion)
+
+// WithBlockSize sets the block size for uploading streams to kusto. Defaults to 8MB.
+func WithBlockSize(size int) Option {
+	return func(s *Ingestion) {
+		s.blockSize = size
 	}
+}
+
+// WithConcurrency sets the concurrency for uploading streams to kusto. Defaults to 50.
+func WithConcurrency(concurrency int) Option {
+	return func(s *Ingestion) {
+		s.concurrency = concurrency
+	}
+}
 
```

Comment:
This is better, so thanks for the changes.

My thinking is that we should not hide the TransferManager.  By just providing two knobs, the user can only turn the two knobs inside a single TransferManager.  We can give them much more control by simply letting them specify the TransferManager they want to use and telling them what the default is.  If they are really concerned about memory, they can simply use the static buffer TM.  They can even create their own that satisfies their needs.

```go
// WithTransferManager provides a custom azblob TransferManager for copying our ingest data to
// Azure Blob Storage for ingestion into Kusto during non-streamed ingestion.  By default 
// we use an azblob.SyncPool that uses 8MiB blocks and 50 concurrent connections.   You can create a
// one with your own settings using azblob.NewSyncPool() or the static allocation with azblob. NewStaticBuffer().
func WithTransferManager(t *azblob.TransferManager) Option {
  return func(s * Ingestion) {
    s.tm = t
  }
}
```

Now in New(), we would do:
```go

func New(client *kusto.Client, db, table string, options ...Option) (*Ingestion, error) {
        mgr, err := getManager(client)
	if err != nil {
		return nil, err
	}

        i := &Ingestion{
		client: client,
		mgr:    mgr,
		db:     db,
		table:  table,
		fs:     fs,
	}

        for _, option := range options {
		option(i)
	}

        if i.tm == nil {
            tm, err := azblob.NewSyncPool(blockSize, concurrency)
	    if err != nil {
		return nil, err
	    }
       }
       ....
```
You would then change filesystem.New() to simply take a TransferManager.  It doesn't need options at this point, as its internal/, so no user can import it.  This code also assumes we move blockSize and concurrency to ingest.go instead of filesystem.go .

Now the users have the full power of the TransferManager and I think it simplifies this down some.

Reply:
@element-of-surprise , I don't like this change.
It exposes to much of the internal implementation to the user.

The previous commit was far better and needed just a little more touch ups (create structs with file system defaults rather then zero, use the size and concurrency in upload File & stream).
@AsafMah , can you revert and push a change with what we discussed?